<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>

<head>

<title>IGB 4.50 Release Notes.</title>

<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />

<meta http-equiv="Content-Language" content="en-US" />

<link rel="stylesheet" type="text/css" media="screen" href="/stylesheet.css" />

<link rel="icon" href="/favicon.ico" />

</head>





<body>

<br>

<h1>IGB 4.50 release notes.</h1>



Highlights:<br><br>



IGB 4.50 adds new capabilities for file parsing and display manipulation.

It now supports additional types of ".chp" file, multi-sequence

".fasta" files, and ".wig" graph files.  It is now easier to

 use the same files in IGB and the UCSC Genome Browser.

IGB can now display scores of annotations through color, and allows

the showing of annotations for a single strand or both strands.<br><br>



<ul>



<li><b>Version 4.50</b>  

<ul>



<li>We have made some changes that allow IGB and the

<a href="http://genome.ucsc.edu/cgi-bin/hgGateway">UCSC Genome Browser</a> 

to share data files more easily.

<ul>

  <li>Parsing of <a href="http://genome.ucsc.edu/google/goldenPath/help/customTrack.html#TRACK">Track Lines</a> is possible

      for formats "BED", "GFF", "PSL" and "WIG", and is consistent with

      the UCSC browser.

  <li>Added support for reading the <a href="http://genome.ucsc.edu/google/goldenPath/help/wiggle.html">Wiggle Graph Format (".wig")</a>

  <li>The range coordinate box in IGB now allows input of coordinates in

      the same format used by UCSC, e.g. "chr1:1,000,000-2,000,000", so

      that you may easily copy and paste coordinates from one program 

      to the other.

</ul>

Each of these points is described in detail below.

<br><br>

</li>





<li>

IGB now allows coloring exons, and other features, based on their score.

<ul>

  <li>Coloring by score can be turned on and off from a menu item.

  <li>For formats that support track lines, turn on coloring by score  

  with the setting <i>useScore=1</i>.

  <li>The colors are interpolated from the given foreground and 

  background colors.

  <li>Scores must range between 1 and 1000.

  <li>Annotations with no score, or with score=0, will be displayed

  in the full foreground color.

</ul>

Examples of using shading to display scored data.  This 

<a href="data/GFF1_example.gff">sample gff file</a> is shown below as it 

displays in IGB and in the UCSC browser.  The file is compatible with

both browsers, but there are some minor differences in the display.

<br><br><img src="images/scores-igb.gif">

<br>Figure: Example of displaying scores by color in IGB.

Note that it is possible to show different scores for each exon

in a single mRNA.<br><br>

<img src="images/scores-ucsc.gif">

<br>Figure: The same file viewed in the UCSC browser.

This browser does not show different scores for each exon; 

the color of the whole mRNA is based on the average score of the exons.

The only allowed score color schemes in the UCSC browser are blue, brown and gray.

Here, for the track where we specified the color red, red is used only

for the lables, and grayscale is used for the scores.<br><br>

</li>





<li><b>Track Lines.</b>

Track lines were designed for use in the UCSC Genome Browser.

We have chosen to allow IGB to also make use of these track lines to

control some of the display properties.  Not all track line properties

are supported in exactly the same way in the two browsers, but the behaviour

is very similar.

<ul>

  <li>Track lines are supported for these file formats in both browsers:

      ".bed", ".wig", ".psl", ".gff".  The line must begin with "track".

   <li>IGB also supports track lines for ".gff3" format.

      In order to comply with the 

      <a href="http://www.sequenceontology.org/gff3.shtml">GFF3 specification</a>,

      the track line in a GFF3 file must begin with "##track". Thus it is a non-standard

      directive, allowed but not defined by the official GFF3 specification.

   <li>IGB makes use of these track-line properties 

      <i>name</i>, <i>color</i>, <i>description</i>, <i>useScore</i>, 

      <i>url</i>  and <i>visibility</i>.

      (Other properties are ignored, but some may be supported 

      in a later version of IGB.)

    <li>In IGB <i>visibility=1</i> or <i>visiblity=dense</i> will cause the

      track to be displayed in "collapsed" mode.  All other settings of

      <i>visibility</i> will cause the track to display in "expanded" mode.

</ul>

<br><br>

</li>



<li><b>Wiggle Format</b>  IGB can now parse graph data in the UCSC

<a href="http://genome.ucsc.edu/google/goldenPath/help/wiggle.html">Wiggle (".wig")</a> 

format.  Below is an example of a <a href="data/wiggleExample.wig">sample wiggle file</a>

displayed in both IGB and the

UCSC browser.

<br><br><img src="images/wiggle_igb.gif">

<br>Figure: Example of displaying a wiggle file in IGB.

<br><br>

<img src="images/wiggle_ucsc.gif">

<br>Figure: The same file viewed in the UCSC browser.

<br><br>

</li>







<li>The <a href="http://biodas.org/documents/das2/das2_protocol.html">DAS/2</a>

client now complies completely with the final

<a href="http://biodas.org/documents/das2/das2_get.html">DAS/2 specification 

for feature data retrieval</a>.<br><br>

</li>



<li>This release adds support for parsing "CHP" files for

expression chips.  The genomic locations of probesets

are not included in these files, but IGB will automatically load

that information as needed from a web server using the

<a href="http://biodas.org/documents/das2/das2_protocol.html">DAS/2</a> protocol.

Affymetrix data formats are described in documents from the

<a href="http://www.affymetrix.com/support/developer/index.affx">Affymetrix

Developers Network</a>.

<br><br>

</li>





<li>IGB now allows configuration of multiple web links from

each feature type.  Links can currently be specified via

track lines in file formats that support track lines, and in DAS XML files.

It is also possible to specify links by regular-expression 

matching in an igb_prefs.xml file.  (Since we plan to replace all of the capabilities

of the igb_prefs.xml file with a graphical interface, the continued use of the

igb_prefs.xml file is discouraged.)<br><br>

<img src="images/web_links.gif"></img><br>

Figure: Multiple web links for an annotation.  Here there are two separate

links for the selected annotation.

<br><br>

</li>



<li>You can now choose to show or hide the "+" and "-" strands of data.

Access this option via <b>View&rarr;Strands</b>.

<br><br><img src="images/show_strands.gif">

<br>Figure: You may choose to hide or show "+", "-" and "+/-" strands.

<br><br></li>



<li>The current position of the hairline is now shown in a new field in the lower-left. 

Viewing the hairline position via <b>View&rarr;Toggle Hairline Label</b> 

also remains possible.

<br><br><img src="images/hairline_position.gif">

<br>Figure: The current location of the hairline is displayed in the lower left.

<br><br>

</li>



<li>You can now load a FASTA formatted file that contains sequence

data for multiple chromosomes (or other sequences).  FASTA format

is recommended only for small genomes.  Open these files in the same way

that you open any other file: <b>File&rarr;Open</b>.<br><br></li>



<li>Bug fixed: The buttons for performing arithmetic operations on two graphs

have been fixed.  (In version 4.48, they all performed subtraction, instead

of the full four operations.)<br><br>

</li>



<li>Bug fixed: When "BAR" format files contain datat from multiple genomes,

the data from all those genomes will be read and made available

through the "Data Access" tab.<br><br></li>

</ul>



</ul>



<div id="footer">

  <p class="center">

  IGB is part of the <a href="http://genoviz.sourceforge.net">GenoViz project.</a>

  Source code for IGB is released under the 

  <a href="http://www.opensource.org/licenses/cpl1.0.php">Common Public License, v1.0</a> an 

  <a href="http://www.opensource.org/docs/definition.php">OSI approved</a>

  open source license. IGB uses other open source software packages

  which are covered by their own open source licenses. </p>

  <p class="center">

  IGB is Copyright (c) 2000-2006 Affymetrix, Inc. Research

  and development of IGB is supported in part by NIH grant

  R01HG003040</p> 

  <p class="center">

  <a href="http://sourceforge.net"><img src="http://sourceforge.net/sflogo.php?group_id=129420&amp;type=1" alt="SourceForge.net" /></a>

  &nbsp;

  &nbsp;

  &nbsp;

  <a href="http://www.opensource.org/docs/definition.php"><img src="http://opensource.org/trademarks/osi-certified/web/osi-certified-60x50.jpg" alt="OSI Certified" width="60" height="50" border="0"></a>

  </p>

</div>

</body>

</html>

